src/backend/distributed/utils/resource_lock.c

```diff
@@ -316,3 +318,84 @@ LockRelationShardResources(List *relationShardList, LOCKMODE lockMode)
 		}
 	}
 }
+
+
+/*
+ * LockParentShardResourceIfPartition acquires a shard resource lock on colocated
+ * shard of parent table of given shard's relation if given shard belongs to a
+ * partition.
+ */
+void
+LockParentShardResourceIfPartition(uint64 shardId, LOCKMODE lockMode)
+{
+	ShardInterval *shardInterval = LoadShardInterval(shardId);
+	Oid relationId = shardInterval->relationId;
+
+	if (PartitionTable(relationId))
+	{
+		int shardIndex = ShardIndex(shardInterval);
+		Oid parentRelationId = PartitionParentOid(relationId);
+		uint64 parentShardId = ColocatedShardIdInRelation(parentRelationId, shardIndex);
+
+		LockShardResource(parentShardId, lockMode);
+	}
+}
+
+
+/*
+ * LockParentRelationShardResourcesIfPartition acquires a shard resource lock on
+ * colocated shard of parent tables of given shards' relation if given shard
+ * belongs to a partition.
+ */
+void
+LockParentRelationShardResourcesIfPartition(List *relationShardList, LOCKMODE lockMode)
+{
+	ListCell *relationShardCell = NULL;
+
+	/* lock shards in a consistent order to prevent deadlock */
+	relationShardList = SortList(relationShardList, CompareRelationShards);
```

Comment:
Hmm, not necessarily. I think we should first extract the co-located shard ids on parent tables, then sort them and acquire lock on them.